doc/developer/platform/architecture-storage.md

```diff
@@ -107,25 +107,142 @@ All commands are durably persisted and their effect survives crashes and reboots
+
+1. Definiteness: this should be provided by `persist`, and isn't meant to be something we can screw up.
+2. "Fault tolerance": the source should produce specific output even in the presence of unspecified i. failures and restarts, ii. concurrency with other source implementations.
+3. "Bounded input reliance": the raw input data we have to keep around to provide these properties should be as limited as possible.
+
+## A recipe for source ingestion
+
+What follows is a recipe for source ingestion that means to provide three properties:
+
+In addition, the "recipe" aspect of this means to make is "simple" to determine that you have these properties.
+You could certainly hand roll something that also has the properties, but then you have a bunch of work to do to convince folks.
+By following the recipe, the goal is that you shouldn't have to convince anyone of anything.
+
+### Ingredients
+
+1. We have some "raw sources" of data.
+
+    These sources implement something like our `SourceReader` trait, which has a method `get_next_message()` that produces an ordered sequence of messages.
+    Messages come with some ordered "address" (e.g. a (part, offset) pair or a LSN position or byte offset).
+    The messages are emitted in an order that matches the address (the address never goes backwards).
+    The source can be initialized at an address.
+    For each address, the source produces the same message to all readers (or "NA" if it's gone).
+
+2. We have access to the `persist` framework that can record evolving collections
+
+    The framework can accept arbitrary `Append` statements and confirm when the are durably recorded.
+    It handles contended writes, picking at most one winner, and providing a single consistent history of updates.
+
+3. We have an evolving understanding of what should happen to our data from raw sources to `persist`.
+
+    We want some amount of flexibility here, as we may be surprised by the requirements of source decoding, or new opportunities.
+    The "understanding" should reflect deterministic computation.
+    Non-determinism is something we can do with an advanced recipe (more `persist` stuff, in the spirit of reclocking, where we record non-deterministic choices).
+
+### The recipe
+
+There is some prep work to do first:
+
+1.  Identify all "raw sources" that contribute to some intended output (e.g. kafka topics, metadata, schema registry).
+
+2.  "Reclock" each input, by choosing and recording in `persist` a collection of source addresses that are monotonically increasing (do not go backwards, but needn't always go forward).
+
+    We should use times that roughly track "real time" so that they align with the times of other collections.
+    The reclocked inputs are going to be treated as if their data "become available" at the first moment for which the data's address is covered by the reclocking collection.
+    Note that this step doesn't use `t` yet.
```

Comment:
This is the first mention of `t`. Did something get reordered, maybe?

Reply:
I moved reclocking up in to "prep work". Should be fixed in the most recent commit.